### PR TITLE
Blog about takewhile & takeUntil

### PR DESCRIPTION
**Pull request by JWO719**

### CODE REVIEW

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+published: false
+title: "TakeWhile, TakeUntil, TakeWhat?"
+cover_image: ""
+description: "The RxJS TakeWhile and TakeUntil operator already sound similar and behave similar. So what's the difference?"
+tags: rxjs, javascript, typescript, webdev
+series:
+canonical_url:
+---
+
+It's already well known, that there are situations where you actually want to unsubscribe from your Observables.
```
Comment (SanderElias) [inline blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: I would rephrase this. You don't "want" to unsubscribe. One MUST unsubscribe unless 100% sure it's not needed.

My personal opinion is that storing and using the subscription is the only safe way, that you can easily explain to new RX users.  There are no "but" or "if" that surround that.

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+There are several
+ways to do this. In several projects, I stumbled across a mixture of using [`takeWhile`](https://rxjs.dev/api/operators/takeWhile) & [`takeUntil`](https://rxjs.dev/api/operators/takeUntil). This raises the question, why are there two of that, which sounds similar and act similarly?
+
+## About TakeWhile
+
+Well, let's take a look at a quick example. The first code snippet we are looking at is using `takeWhile` to unsubscribe from an observable. I used an Angular Component here, to couple the lifecycle of the observable to the lifecycle of the component.
+Short disclaimer: It doesn't necessarily need to be an Angular component, it's just the framework I'm the most familiar with.
```
Comment (SanderElias) [inline blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: Using angular code will make this an angular article in the eyes of most readers. I understand this is the easiest to come up with an actual example,  but please consider making it rx code purely.

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md (hunk 1 above)
Comment (tobmaster) [reply blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: I would try to avoid implications like "It's already well known". Instead maybe use something like "It might be know to you already that..."

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md (hunk 2 above)
Comment (tobmaster) [reply blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: Maybe add a small sentence in an extra box that tells non angular users what ngOnInit and ngondestroy do to understand the case here

Comment (JordyLangen) [reply blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: Or rather use a plain JavaScript example? Maybe using something like `mousedown` or `mousemove` events which you don't want anymore after release the mouse click (`mouseup`). Using "dragging an html element" or "drawing a line" as an example.

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+So all in all both code snippets look similar and behave similar, right? Well, No!
+Let's have a look at the marble diagrams describing those operators because this will highlight the difference between those two operators.
+
+![takeUntil marble diagram, kindly provided by Michael Hladky](./assets/takeUntil.png)
+_takeUntil marble diagram, kindly provided by [Michael Hladky](https://twitter.com/Michael_Hladky)_
+
+![takeWhile marble diagram, kindly provided by Michael Hladky](./assets/takeWhile.png)
+_takeWhile marble diagram, kindly provided by [Michael Hladky](https://twitter.com/Michael_Hladky)_
+
+The problem here is that `takeWhile` is intended to take an incoming notification, use some comparison on it, which might lead to an unsubscribe. The important fact is, that `takeWhile` is triggered by the incoming notification and might unsubscribe afterwards. `TakeUntil` in contrast is triggered by the passed Observable.
```
Comment (JordyLangen) [inline blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: >use some comparison on it
I would slightly rephrase this. It takes an incoming notification, and check a specified condition. It does not necessarily need to be a comparison.

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+longLivingObservable$
+  .pipe(
+    tap(() => this.startAnimation()),
+    switchMap(val => this.makeHttpCall(val)),
+    takeWhile(val => this.alive),
+  )
+  .subscribe();
+```
+
+So what's the issue with this piece of code? Well, our component is already destroyed and due to the needed notification, needed before unsubscribing kicks in, we will start an animation and trigger an HTTP call. This is probably unwanted and just afterwards we will check if we want to unsubscribe from our Observable. Besides the fact that those operations are totally superfluous, it also might break our app or pollute our state.
```
Comment (JordyLangen) [inline blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: > So what's the issue with this piece of code? Well, our component is already destroyed and due to the needed notification, needed before unsubscribing kicks in,

This last part (`needed before unsubscribing kicks in`) reads kind of weird. I suppose you mean that is problematic because the `takeWhile` is at the end of th chain, meaning that it will still trigger the animation and http call?

Comment (JordyLangen) [reply blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: You could even say that it would cause a memory leak if the the `longLivingObservable$` never emits (nor completes) after the component is destroyed

File: blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+_Now maybe one might argue: "Well, I could move the `takeWhile` operator at the very beginning of the observable pipeline!"_
+
+That's true, you could do this, and you will save the unneeded operations, which is a good start, but you won't unsubscribe from inner Observables. So if the Observable returned by `makeHttpCall` is a long-living Observable, it will not unsubscribe from that, if the `takeWhile` is before `switchMap` in the pipe. By the way, the same goes for `takeUntil`, so make sure to have the unsubscribe-operator at the very end of your pipe.
+
+## The Solution
+
+Don't get me wrong, `takeWhile` is an amazing operator, but just if you actually use the incoming value to determine, whether you want to unsubscribe or not! Do not depend on "global" state, when using `takeWhile`.
```
Comment (JordyLangen) [inline blog-posts/takeuntil-takewhile/takeWhile-takeUntil-takeWhat.md]: Would it make sense to add an real world example using `takeWhile` to strengthen your point and show the difference in usage between `takeUntil` and `takeWhile`?
